include/nlohmann/json.hpp

```diff
@@ -2237,6 +2237,10 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
 
             return default_value;
         }
+        else if (JSON_HEDLEY_LIKELY(is_null()))
```

Comment:
Please don't change the behavior of the `value` function. This would break the API.

Reply:
All the PR is about this change.
Which API may be broken?

Reply:
No. The issue is that when `NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT` is applied to an empty JSON value, there is a problem. You changed the behavior of the `value` function which may be called by the macro, but is also a public API function.

From its documentation (https://json.nlohmann.me/api/basic_json/value/#parameters):

> [The function throws] [type_error.306](https://json.nlohmann.me/home/exceptions/#jsonexceptiontype_error306) if the JSON value is not an object; in that case, using value() with a key makes no sense.

Reply:
Maybe, but that's still changing the behavior of classes using these macros, which were introduced in January 2022, and released in 3.11.0 in August of 2022, so they've been part of 4 releases over the last 2+ years.

```
It will not throw an exception in `from_json()` due to a missing value in the JSON object, but can throw due to a mismatched type. The `from_json()` function default constructs an object and uses its values as the defaults when calling the `value()` function.
```

Reply:
Oh, this isn't the error that I thought it was.  I thought it was `from_json` on trying to get `a` and `b` from `{ "a": null, "b": null }`, not `{}`.  This would actually work properly on `{ "c": "foo" }`.

In this case, I think it would be okay for `from_json` to check for the root `json` being `null` instead of an object.

Reply:
```
#define NLOHMANN_JSON_FROM_WITH_DEFAULT(v1) nlohmann_json_t.v1 = nlohmann_json_j.value(#v1, nlohmann_json_default_obj.v1);
```
Maybe this, which matches the current documented behavior?
```
#define NLOHMANN_JSON_FROM_WITH_DEFAULT(v1) nlohmann_json_t.v1 = !nlohmann_json_j.is_null() ? nlohmann_json_j.value(#v1, nlohmann_json_default_obj.v1) : nlohmann_json_default_obj.v1;
```

Reply:
yes, this is the idea to follow

Reply:
Okay, so how did you actually get to this condition?  Is there some reason that you can't check that the `json` object is empty first before creating `Foo` from it?  I assume it's part of a larger document and object model.  I'm wondering if there is a higher level where this could be avoided.

Reply:
I get to this condition while deserialazing a wrong json content.
In this case the json instance is discarded, so instead we wanted to use an empty one with `json = {}`
then we get an exception instead of getting default values in Foo constructor

so: using a json instance with bar content instead of foo content is working,
but: using a json instance empty without bar content nor foo content is throwing exception

IMHO: json = {} is a valid and legal constructor

```
static void JsonTest_parse(std::string content) {
  // with an empty_content or an invalid_content, json is discarded
  auto json = nlohmann::json::parse(content, nullptr, false);
  if (json.is_discarded()) {
    LOG_ERROR("Failed to parse: {}", content);
    // here wanted to build a new empty json
    json = {};  // set custom config to empty
  }
  // an exception is thrown instead of getting default values
  auto foo = Foo(json);
  foo.log();
};

// with an empty json, an exception is thrown instead of getting default values
static void JsonTest_empty_json() {

  // this is OK to parse a content with bar instead of foo
  std::string valid_content{R"({"bar": true})"};
  JsonTest_parse(valid_content);

  // both tests empty_content and invalid_content are discarded
  // and can't build Foo with its default values
  std::string empty_content{""};
  JsonTest_parse(empty_content);

  std::string invalid_content{R"({xxx "bar": true})"};
  JsonTest_parse(invalid_content);
}
```

Reply:
Is there some reason that you can't just create the default object instead of constructing it from an empty json object?  I assume that the code you shared is just testing this particular issue, and is not representative of your actual code.  If it is representative, you can just do this:

```
static void JsonTest_parse(std::string content) {
  // with an empty_content or an invalid_content, json is discarded
  auto json = nlohmann::json::parse(content, nullptr, false);

  Foo foo;
  if (!json.is_discarded()) {
    foo = json;
  } else {
    LOG_ERROR("Failed to parse: {}", content);
  }
  foo.log();
};
```

Reply:
yes we can, and there are plenty of options to solve this, but:

> API is like a joke. If you have to explain it, it’s not that good.

`NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT` should means that if the key is not found during deserialization the defaults will be used.

Why artificially limit it? This adds complexity for the API users while not providing any clear benefits

Reply:
That's exactly what it means, but it currently has a precondition that the place where you are looking for the key is actually something that can contain a key.  Now, it's at least true here that this is just looking to handle it for a `null`, which has no data, so it's reasonable to make that change to relax the precondition a bit.

What I'm wondering, however, is if your actual code is doing exactly what you are showing here, and intentionally creating an empty `json` object and manually using it to create the `Foo` object, in which case it seems like the best way to fix that immediately is to not do that.

If this is just reproducing a smaller part of a larger system where you are getting a `null` instead of an `object` in a larger JSON string, and assigning it to `Foo` inside a larger data structure, then there may be something to be done even one level higher in your code so that you never even hit trying to create `Foo` from the empty `json`.

Both of those fixes are more immediately applicable to your code than updating this library and waiting for a new release.

Reply:
Yes we made the necessary changes in our code to not try to use an empty JSON.
For the library, we should make this change, would you agree?

Reply:
I would need to see what's changed in the end. If this is about fixing the behavior of a macro, then I'm in. If it's about changing the behavior of an existing public API function, then I am out.

Reply:
Yes, just this change to the macro:
https://github.com/nlohmann/json/pull/4508#discussion_r1852637742

```
#define NLOHMANN_JSON_FROM_WITH_DEFAULT(v1)
```

From
```
nlohmann_json_t.v1 = nlohmann_json_j.value(#v1, nlohmann_json_default_obj.v1);
```

To
```
nlohmann_json_t.v1 = !nlohmann_json_j.is_null() ? nlohmann_json_j.value(#v1, nlohmann_json_default_obj.v1) : nlohmann_json_default_obj.v1;
```

Reply:
only this macro as https://github.com/nlohmann/json/pull/4508#discussion_r1852637742

```
#define NLOHMANN_JSON_FROM_WITH_DEFAULT(v1) nlohmann_json_t.v1 = !nlohmann_json_j.is_null() ? nlohmann_json_j.value(#v1, nlohmann_json_default_obj.v1) : nlohmann_json_default_obj.v1;
```